parser/parser.go

```diff
@@ -2,169 +2,208 @@ package parser
 	m := message.Params{}
 	err := json.Unmarshal(params, &m)
-	return m, err
-}
-
-func ParseParamsIncludingMessage(params json.RawMessage) (message.ParamsIncludingMessage, error) {
-	m := message.ParamsIncludingMessage{}
-	err := json.Unmarshal(params, &m)
-	if err != nil {
-		return m, lib.ErrEncodingFault
+	//L3Jvb3Q= is b64 for "/root"
+	decodedChannel, err := lib.Decode(m.Channel)
+	if !strings.HasPrefix(string(decodedChannel), "/root") || err != nil {
+		log.Printf("channel id doesn't start with /root but is %v", m.Channel)
+		return m, lib.ErrRequestDataInvalid
 	}
-	d, err := lib.Decode(m.Channel)
-	m.Channel = string(d)
 	return m, err
 }
 
+// ParseMessage parses a json.RawMessage into a message.Message structure.
 func ParseMessage(msg json.RawMessage) (message.Message, error) {
 	m := message.Message{}
 	err := json.Unmarshal(msg, &m)
-
-	d, err := lib.Decode(m.Sender)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Sender = string(d)
-
-	d, err = lib.Decode(m.MessageId)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.MessageId = string(d)
-
-	d, err = lib.Decode(m.Signature)
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Signature = string(d)
-
-	//Marshall automatically encode Json.rawMessage in Base64
-	d, err = lib.Decode(string(m.Data))
-	if err != nil {
-		return m, lib.ErrEncodingFault
-	}
-	m.Data = d
-
-	for i := 0; i < len(m.WitnessSignatures); i++ {
-		d, err = lib.Decode(m.WitnessSignatures[i])
-		if err != nil {
-			return m, lib.ErrEncodingFault
-		}
-		m.WitnessSignatures[i] = string(d)
-	}
 	return m, err
 }
-func ParseWitnessSignature(witnessSignatures string) (message.ItemWitnessSignatures, error) {
+
+// ParseWitnessSignature parses a json.RawMessage into a message.ItemWitnessSignatures structure.
+func ParseWitnessSignature(witnessSignatures json.RawMessage) (message.ItemWitnessSignatures, error) {
 	m := message.ItemWitnessSignatures{}
-	err := json.Unmarshal([]byte(witnessSignatures), &m)
+	err := json.Unmarshal(witnessSignatures, &m)
+	return m, err
+}
 
-	d, err := lib.Decode(m.Signature)
-	if err != nil {
-		return m, lib.ErrEncodingFault
+// ParseData parses a json.RawMessage into a message.Data structure. It calls dataConstAreValid to check restrictions
+// defined in the jsonRPC protocol.
+func ParseData(data string) (message.Data, error) {
+	m := message.Data{}
+	err := json.Unmarshal([]byte(data), &m)
+	if dataConstAreValid(m) {
+		return m, err
+	} else {
+		return m, lib.ErrRequestDataInvalid
 	}
-	m.Signature = string(d)
+}
 
-	d, err = lib.Decode(m.Witness)
-	if err != nil {
-		return m, lib.ErrEncodingFault
+// dataConstAreValid verifies following constraints :
+// * object is one of "lao", "message", "meeting"
+// * action is one of "create", update_properties", "state", "witness"
+// * creation and last modified are positive integer
+func dataConstAreValid(m message.Data) bool {
+	switch m["object"] {
+	case "lao", "message", "meeting":
```

Comment:
It is crucial to update this list of cases and the one two lines below to match the most up-to-date protocol specs

Reply:
yep, an IMPORTANT notice is in the comment of the next PR